### ISSUE TITLE
Qwik: valiForm in formAction doesn't apply pipeline transformations

### ISSUE THREAD

**Issue by MaieulChevalier** [fabian-hiller/modular-forms#132]
I'm trying to use drizzle-valibot package and I have to use valibot's parse function in my formAction in order to apply pipeline transformations (like `toTrimmed()`, etc.) because valiForm$ doesn't.

### What I do

```typescript
export const useFormAction = formAction$<UpdateProfileForm>(async (values, event) => {
  const session: Session | null = event.sharedMap.get("session");

  try {
    // Necessary because valiForm doesn't apply transformations
    const parsedValues = parse(requestSchema, values);
    await db
      .update(users)
      .set(parsedValues)
      .where(eq(users.name, session?.user.name as string));
  } catch (error) {
    console.log(error);
  }
}, valiForm$(requestSchema));
```

### What I was expecting to work out of the box
```typescript
export const useFormAction = formAction$<UpdateProfileForm>(async (values, event) => {
  const session: Session | null = event.sharedMap.get("session");

    await db
      .update(users)
      .set(values) // values would have correct transformations
      .where(eq(users.name, session?.user.name as string));

}, valiForm$(requestSchema));
```

### Reason
Looking at the code I suspect this to be an issue with the way valibot is called in modular-form's code with `_parse(values, { abortPipeEarly: true })`. I guess the problem comes from `{ abortPipeEarly: true }`.

### Potential solutions
I guess the `{ abortPipeEarly: true })` argument isn't there for nothing, but I see zod package uses `safeParse`. Can't valiForm be similarly written with `parse` or `safeParse` function?

**Comment by fabian-hiller** [comment]
The "problem" is related to the fact that Modular Forms uses Valibot only for validation and does not consider the output. The background is that with `transform` it is possible to make more extensive changes to the data set and Modular Forms does not know how to handle this. For example, if the data type changes, Modular Forms could break.

More infos:

- https://github.com/fabian-hiller/modular-forms/issues/20
- https://github.com/fabian-hiller/modular-forms/issues/40

For now, you can use Modular Forms's `transform` feature to trim a string: https://modularforms.dev/qwik/guides/transform-inputs

**Comment by MaieulChevalier** [comment]
Thanks for pointing to the related issues.

I'll try to create a similar function as proposed in #20.

> For now, you can use Modular Forms's transform feature to trim a string: https://modularforms.dev/qwik/guides/transform-inputs
The problem with this approach is that it is not the DX I'm looking for. I want to define all my schema in one place and use it throughout my application. Using the `transform` feature of modular form, I will basically be rewriting the same code twice.

**Comment by MaieulChevalier** [comment]
But I wonder if adding `const parsedValues = parse(requestSchema, values)` in my action isn't just the best API I can use right now.

**Comment by fabian-hiller** [comment]
> But I wonder if adding `const parsedValues = parse(requestSchema, values)` in my action isn't just the best API I can use right now.

Yes, this is the best workaround at the moment.

**Comment by fabian-hiller** [comment]
Thank you! Yes I will give some thought to this. Maybe it can make sense to change Modular Forms or to develop a form library with a similar API than Modular Forms that is perfectly tailored to Valibot.

**Comment by maiieul** [comment]
I don't think Modular Form's current API is so far from Zod/Valibot that it necessitates a new form library. I think the underlying issue here is mainly server-side; since applying transformations at the form level could yield to unexpected/buggy behavior client-side.

### Server-side

With server actions, I think it makes sense to always apply validation (as is currently the case) **and**  transformations (as is not currently the case) to respect what the schema says.

In my schema, I have

```typescript
  firstName: () =>
    string([
      toTrimmed(),
    ]),
```
But `toTrimmed` doesn't get applied in my `formAction$` unless I use my own parsed values with `const parsedValues = parse(requestSchema, values)` as explained above.

I expected this to work out of the box by applying `toTrimmed` with `valiForm$` as the second argument to `formAction$`. Without that transformation getting applied, I will have strings like "blablabla       " (with many spaces) in my database, which is something I definitely want to avoid.

So in

```typescript
export const useFormAction = formAction$<LoginForm>((values) => {
  // Runs on server
}, valiForm$(LoginSchema));
```

The second argument with `valiForm$(LoginSchema)` should apply transformations out of the box, and even transformations that change the data type shouldn't be of issue here (right?). `valiForm$` might not be easy to modify due to client-side requirements (see below :point_down:), so perhaps we could introduce a new `valiParse$` function?

### Client-side

On the client, I think that it doesn't always make sense to apply transformations client side. 

For example, a `toTrimmed` transformation could prevent the user from adding multiple spaces at once client-side, and therefore tell him that it can't be done. On the other hand, what would `transform(number(), (amount) => amount * 100)` mean? Every time the user enters a number, the transformation would add two zeros at the end of it? That wouldn't mean anything.

So I think for client-side transformations, it's best to use Field-level transformations as in https://modularforms.dev/qwik/guides/transform-inputs.

**Comment by itsUndefined** [comment]
Yes. You are correct that the transformations only make sense on SubmitHandler and on formAction$

**Comment by fabian-hiller** [comment]
Thank you for the detailed feedback. I agree in principle and will give it some thought.
